File: ios/BatteryHolder/Services/Cloud/FirmwareRepository.swift
import Foundation
import CryptoKit

enum FirmwareRepositoryError: LocalizedError {
    case http(Int), noDownloadURL, checksumMismatch
    var errorDescription: String? {
        switch self {
        case .http(let code): return "Server returned HTTP \(code)."
        case .noDownloadURL: return "No download URL was provided for this build."
        case .checksumMismatch: return "Downloaded firmware failed its SHA-256 check."
        }
    }
}

/// REST client for the AWS firmware catalog (see docs/AWS_BACKEND.md).
final class FirmwareRepository {
    private let baseURL: URL
    private let session: URLSession
    /// Supplies a Cognito JWT for authorized requests. Returns nil when unauthenticated.
    var tokenProvider: () async -> String? = { nil }

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    private func decoder() -> JSONDecoder {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }

    private func authorized(_ url: URL, method: String = "GET") async -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method
        if let token = await tokenProvider() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    /// List builds available for a board, newest first.
    func listFirmware(boardId: String) async throws -> [FirmwareImage] {
        struct Envelope: Decodable { let items: [FirmwareImage] }
        let url = baseURL.appendingPathComponent("boards/\(boardId)/firmware")
        let (data, response) = try await session.data(for: await authorized(url))
        try Self.check(response)
        return try decoder().decode(Envelope.self, from: data).items
    }

    /// Fetch build detail including a presigned `downloadUrl`.
    func detail(buildId: String) async throws -> FirmwareImage {
        let url = baseURL.appendingPathComponent("firmware/\(buildId)")
        let (data, response) = try await session.data(for: await authorized(url))
        try Self.check(response)
        return try decoder().decode(FirmwareImage.self, from: data)
    }

    /// Download the firmware binary and verify its SHA-256 before returning it.
    func download(_ image: FirmwareImage) async throws -> Data {
        let resolved = image.downloadUrl != nil ? image : try await detail(buildId: image.buildId)
        guard let url = resolved.downloadUrl else { throw FirmwareRepositoryError.noDownloadURL }

        let (data, response) = try await session.data(from: url)
        try Self.check(response)

        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard digest.caseInsensitiveCompare(resolved.sha256) == .orderedSame else {
            throw FirmwareRepositoryError.checksumMismatch
        }
        return data
    }

    private static func check(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else {
            throw FirmwareRepositoryError.http(http.statusCode)
        }
    }
}

File: android_app/lib/services/firmware_repository.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;

import '../models/firmware_image.dart';

class FirmwareRepositoryException implements Exception {
  final String message;
  const FirmwareRepositoryException(this.message);

  factory FirmwareRepositoryException.http(int code) =>
      FirmwareRepositoryException('Server returned HTTP $code.');

  static const noDownloadURL = FirmwareRepositoryException(
      'No download URL was provided for this build.');
  static const checksumMismatch = FirmwareRepositoryException(
      'Downloaded firmware failed its SHA-256 check.');

  @override
  String toString() => message;
}

/// REST client for the AWS firmware catalog (see docs/AWS_BACKEND.md).
class FirmwareRepository {
  final Uri baseURL;
  final http.Client _client;

  /// Supplies a Cognito JWT for authorized requests. Returns null when
  /// unauthenticated.
  Future<String?> Function() tokenProvider = () async => null;

  FirmwareRepository({required this.baseURL, http.Client? client})
      : _client = client ?? http.Client();

  Future<Map<String, String>> _headers() async {
    final token = await tokenProvider();
    return token == null ? {} : {'Authorization': 'Bearer $token'};
  }

  /// List builds available for a board, newest first.
  Future<List<FirmwareImage>> listFirmware(String boardId) async {
    final url = baseURL.resolve('boards/$boardId/firmware');
    final response = await _client.get(url, headers: await _headers());
    _check(response);
    final envelope = jsonDecode(response.body) as Map<String, dynamic>;
    return ((envelope['items'] as List?) ?? const [])
        .map((e) => FirmwareImage.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Fetch build detail including a presigned `downloadUrl`.
  Future<FirmwareImage> detail(String buildId) async {
    final url = baseURL.resolve('firmware/$buildId');
    final response = await _client.get(url, headers: await _headers());
    _check(response);
    return FirmwareImage.fromJson(
        jsonDecode(response.body) as Map<String, dynamic>);
  }

  /// Download the firmware binary and verify its SHA-256 before returning it.
  Future<Uint8List> download(FirmwareImage image) async {
    final resolved =
        image.downloadUrl != null ? image : await detail(image.buildId);
    final url = resolved.downloadUrl;
    if (url == null) throw FirmwareRepositoryException.noDownloadURL;

    final response = await _client.get(url);
    _check(response);

    final digest = sha256.convert(response.bodyBytes).toString();
    if (digest.toLowerCase() != resolved.sha256.toLowerCase()) {
      throw FirmwareRepositoryException.checksumMismatch;
    }
    return response.bodyBytes;
  }

  void _check(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw FirmwareRepositoryException.http(response.statusCode);
    }
  }

  void dispose() => _client.close();
}
